Network interface MAC address set/get

artokin opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->

<!-- 
    Please provide the following information: 

    Description of the the change (what is this fixing / adding / removing?).

    Why the change is needed (if this is fixing a reported issue please summarize what
    the issue is and add the reference. E.g. Fixes #17119).

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
    
-->

Add method set_mac_address that Application can use to set MAC address to the interface.
Update get_mac_address to return MAC address correctly from EMAC.

This PR is copied from: #13902

#### Impact of changes <!-- Optional -->
<!-- 
    If there are any implications for users taking this change then they must be 
    provided here. For Major PR types this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

#### Migration actions required <!-- Optional -->
<!-- 
    This should only be applicable in Major PR types for which this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

### Documentation <!-- Required -->

<!-- 
    Please provide details of any document updates required, including links to any
    related PRs against the docs repository.
    If no document updates are required please specify 'None', this at least tells us
    that this has been considered.
-->

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [x] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [] No Tests required for this change (E.g docs only update)
    [x] Covered by existing mbed-os tests (Greentea or Unittest)
    [] Tests / results supplied as part of this PR
    
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->
@mikter , @mikaleppanen , @juhhei01 , @JarkkoPaso 

<!--
    Request additional reviewers with @username or @team
-->

----------------------------------------------------------------------------------------------------------------

Commit:
Method for adding network interface MAC address

Add method set_mac_address to set network interface MAC address.

Commit:
MAC address set/get support for EMAC interface

Add support to set and get MAC address from EMAC interface.

Commit:
Review correction: Remove unnecessary method

Remove unnecessary method: add_ethernet_interface_ns

Commit:
Allow method override

Remove final keyword to allow inheritance.

Commit:
Update set_mac_address signature

Use nsapi_size_t instread os size_t.

Commit:
Update unittests

paul-szczepanek-arm commented on connectivity/netsocket/include/netsocket/NetworkInterface.h:
```diff
@@ -101,6 +101,24 @@ class NetworkInterface: public DNS {
+     *  All interfaces are not supporting MAC address set. A call to connect()
+     *  will fail if MAC address is provided but not possible to use.
+     *
+     *  @param          mac_addr Buffer containing the MAC address
```
in what format?

artokin replied:
added "in hexadecimal format"

paul-szczepanek-arm commented on connectivity/netsocket/include/netsocket/NetworkInterface.h:
```diff
@@ -101,6 +101,24 @@ class NetworkInterface: public DNS {
+     *  set before calling the interface connect() method. EUI-48 MAC addresses
+     *  are used for Ethernet while Mesh interface is using 8-byte EUI-64 address.
+     *
+     *  All interfaces are not supporting MAC address set. A call to connect()
```
can't the set_mac_address return an error if it's not supported?

artokin replied:
Not easily as the interface initialisation happens during the connect()-method call.

mikter replied:
If connect fails I think this makes quite inflexible interface as you can't recover from this and can't make a feature to read Mac from Factory tool, but for devices that has HW support for getting MAC it should override the factory tool configuration. So connect should not fail, but trace out an warning

artokin replied:
@mikter , good point!
Connect() failure is now ignored in the correction commit. Would you please review again?

paul-szczepanek-arm commented on connectivity/netsocket/include/netsocket/NetworkInterface.h:
```diff
@@ -101,6 +101,24 @@ class NetworkInterface: public DNS {
 
+    /** Set the MAC address to the interface.
+     *
+     *  Provided MAC address is set the the network interface. Address must be
```
```suggestion
     *  Provided MAC address is set on the network interface. Address must be
```

paul-szczepanek-arm commented on connectivity/netsocket/tests/TESTS/network/emac/emac_TestNetworkStack.cpp:
```diff
@@ -155,6 +155,11 @@ char *EmacTestNetworkStack::Interface::get_mac_address(char *buf, nsapi_size_t b
 
+nsapi_error_t EmacTestNetworkStack::Interface::set_mac_address(uint8_t *buf, nsapi_size_t buflen)
+{
+    return NSAPI_ERROR_OK;
```
Shouldn't this be NSAPI_ERROR_UNSUPPORTED?

Commit:
Review corrections

-Update method set_mac_address description
-Update EMAC test method return value

Commit:
Update EMAC connect logic with MAC address change

Do not fail EMAC interface during connect if MAC address change is
not supported in the selected interface.

Commit:
Follow Mbed OS coding style

Commit:
Fix method description

Commit:
Fix unittests

Use path in "mbed-trace/mbed_trace.h" to keep unittest happy.

pan- commented on connectivity/netsocket/tests/TESTS/network/emac/emac_TestNetworkStack.h:
```diff
@@ -84,7 +84,9 @@ class EmacTestNetworkStack : public OnboardNetworkStack, private mbed::NonCopyab
          * @return              MAC address as "V:W:X:Y:Z"
          */
-        virtual char *get_mac_address(char *buf, nsapi_size_t buflen);
+        //virtual char *get_mac_address(char *buf, nsapi_size_t buflen);
```
This function is still defined in the implementation file isn't it: https://github.com/ARMmbed/mbed-os/blob/40c038555dc191100231aa256181a302cc983e43/connectivity/netsocket/tests/TESTS/network/emac/emac_TestNetworkStack.cpp#L153 ?

artokin replied:
Absolutely, thanks for noticing this! Fixed in the following commit.

Commit:
Update EmacTestNetworkStack

-Uncomment get_mac_address
-Document set_mac_address

Commit:
Remove API changes

-Remove Nanostack::add_ethernet_interface API change
-Add get_mac_address to MeshEthernetInterface
-Fix compiler warnings by adding overrides

Commit:
Update WICED binary for ARMC6

A new method has been added to the NetworkInterface. Therfore WICED
library needs to be rebuild for ARMC6.
